### PR TITLE
Add a unified TLS stream type (fixes #46)

### LINKED ISSUE #46: Add a bidirectional `TlsStream` wrapper
For some projects, it can be convenient to have a single type that abstracts over client and server versions of the `TlsStream`. This also makes integrating TLS support easier, since `TcpStream` is bidirectional, but there are two different `TlsStream` types in this project.

The code below seems useful in making this a little more ergonomic, at a supposedly very minor performance cost (could apply some `#[inline]` directives to make it even better):

```rust
enum TlsStream<T> {
    Client(TlsClientStream<T>),
    Server(TlsServerStream<T>),
}

impl<T> TlsStream<T> {
    fn tcp(&self) -> &T {
        use TlsStream::*;
        match self {
            Client(io) => io.get_ref().0,
            Server(io) => io.get_ref().0,
        }
    }
}

impl<T> From<TlsClientStream<T>> for TlsStream<T> {
    fn from(s: TlsClientStream<T>) -> Self {
        Self::Client(s)
    }
}

impl<T> From<TlsServerStream<T>> for TlsStream<T> {
    fn from(s: TlsServerStream<T>) -> Self {
        Self::Server(s)
    }
}

impl<T> io::Read for TlsStream<T>
where
    T: AsyncRead + AsyncWrite + io::Read,
{
    fn read(&mut self, buf: &mut [u8]) -> Result<usize, io::Error> {
        use TlsStream::*;
        match self {
            Client(io) => io.read(buf),
            Server(io) => io.read(buf),
        }
    }
}

impl<T> io::Write for TlsStream<T>
where
    T: AsyncRead + AsyncWrite + io::Write,
{
    fn write(&mut self, buf: &[u8]) -> Result<usize, io::Error> {
        use TlsStream::*;
        match self {
            Client(io) => io.write(buf),
            Server(io) => io.write(buf),
        }
    }

    fn flush(&mut self) -> Result<(), io::Error> {
        use TlsStream::*;
        match self {
            Client(io) => io.flush(),
            Server(io) => io.flush(),
        }
    }
}

impl<T> AsyncRead for TlsStream<T> where T: AsyncRead + AsyncWrite {}

impl<T> AsyncWrite for TlsStream<T>
where
    T: AsyncRead + AsyncWrite,
{
    fn shutdown(&mut self) -> Poll<(), io::Error> {
        use TlsStream::*;
        match self {
            Client(io) => io.shutdown(),
            Server(io) => io.shutdown(),
        }
    }
}
```

### PR DESCRIPTION
**Pull request by djc**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -216,6 +216,90 @@ impl<IO: AsyncRead + AsyncWrite> Future for Accept<IO> {
...
+impl<T> TlsStream<T> {
+    pub fn get_io_ref(&self) -> &T {
+        use TlsStream::*;
+        match self {
+            Client(io) => io.get_ref().0,
+            Server(io) => io.get_ref().0,
+        }
+    }
+
+    pub fn get_io_mut(&mut self) -> &T {
```
Comment (quininer) [inline src/lib.rs]:
```suggestion
    pub fn get_io_mut(&mut self) -> &mut T {
```

File: src/lib.rs [hunk 2]
```diff
@@ -216,6 +216,90 @@ impl<IO: AsyncRead + AsyncWrite> Future for Accept<IO> {
     }
 }
 
+pub enum TlsStream<T> {
+    Client(client::TlsStream<T>),
+    Server(server::TlsStream<T>),
+}
+
+impl<T> TlsStream<T> {
```
Comment (quininer) [inline src/lib.rs:228]: We can also add `get_session` and `get_session_mut`.

```rust
fn get_session(&self) -> &dyn Session;

fn get_session_mut(&mut self) -> &mut dyn Session;
```

**Review by quininer — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 1 above)
Comment (djc) [reply src/lib.rs]: Did a slightly different take, have a look.

File: src/lib.rs (hunk 2 above)
Comment (djc) [reply src/lib.rs:228]: Ahh, trait objects. Should have thought of that.

**Review by quininer — APPROVED** [review]
LGTM, Thanks you!

**Comment by djc** [comment]
Would you mind pushing out 0.10.1 to crates.io? Then I can remove this code from my downstream project. :)

**Comment by quininer** [comment]
`0.10.1` has been released!

**Comment by djc** [comment]
Thank you!

**Comment by djc** [comment]
FYI, I just noticed on someone else's machine that this inadvertently introduced a dependency on Rust 1.37.0 due to the use of `Self::Client` and `Self::Server` for enum variants. Not sure if you have MSRV policy or anything, but just so you're aware.

**Comment by quininer** [comment]
I don't have an MSRV policy, but I don't mind supporting more versions of rustc.

**Pull request merged**